### PR TITLE
Adding Together API to the providers

### PR DESCRIPTION
**Pull request by abirharrasse**
We make it possible for users of MetaGPT to use their together api and get access to the agentic features offered by MetaGPT.

### CODE REVIEW

File: metagpt/__init__.py [hunk 1]
```diff
@@ -5,3 +5,34 @@
 # @File    : __init__.py
 
 from metagpt import _compat as _  # noqa: F401
+from metagpt.provider.google_gemini_api import GeminiLLM
```
Comment (better629) [inline metagpt/__init__.py:8]: these codes are already in `metagpt/provider/__init__.py`

File: metagpt/configs/llm_config.py [hunk 2]
```diff
@@ -10,9 +10,9 @@
 
 from pydantic import field_validator
 
-from metagpt.const import CONFIG_ROOT, LLM_API_TIMEOUT, METAGPT_ROOT
+from metagpt.const import LLM_API_TIMEOUT
 from metagpt.utils.yaml_model import YamlModel
-
+from metagpt.const import METAGPT_ROOT, CONFIG_ROOT
```
Comment (better629) [inline metagpt/configs/llm_config.py:15]: use pre-commit to format code

File: metagpt/configs/llm_config.py [hunk 3]
```diff
@@ -97,13 +97,12 @@ def check_llm_key(cls, v):
...
-                raise ValueError(
-                    f"Please set your API key in {root_config_path}. If you also set your config in {repo_config_path}, \nthe former will overwrite the latter. This may cause unexpected result.\n"
-                )
+                 raise ValueError(
+                    f"Please set your API key in {root_config_path}. If you also set your config in {repo_config_path}, \nthe former will overwrite the latter. This may cause unexpected result.\n")
             elif repo_config_path.exists():
                 raise ValueError(f"Please set your API key in {repo_config_path}")
             else:
-                raise ValueError("Please set your API key in config2.yaml")
+                raise ValueError(f"Please set your API key in config2.yaml")
```
Comment (better629) [inline metagpt/configs/llm_config.py:105]: why add f?

File: metagpt/provider/together.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
+import asyncio
+from typing import Optional, Union, AsyncGenerator
+from together import Together
```
Comment (better629) [inline metagpt/provider/together.py:3]: doesn't add together version in requirments.txt

File: metagpt/provider/together.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        response = await asyncio.to_thread(self.client.chat.completions.create, **kwargs)
+        
+        collected_messages = []
+        current_sentence = ""
+        for chunk in response:
+            if chunk.choices:
+                chunk_message = chunk.choices[0].delta.content
+                if chunk_message:
+                    current_sentence += chunk_message
+                    if chunk_message.endswith(('.', '!', '?', '\n')):
```
Comment (better629) [inline metagpt/provider/together.py:55]: why deal with there special chars?

File: metagpt/provider/together.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                        collected_messages.append(current_sentence)
+                        yield current_sentence
+                        print(current_sentence, end='', flush=True)  # Print each sentence as it's completed
```
Comment (better629) [inline metagpt/provider/together.py:58]: use log_llm_stream instead of print

File: metagpt/provider/together.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+from metagpt.configs.llm_config import LLMConfig, LLMType
+from metagpt.provider.llm_provider_registry import register_provider
+from metagpt.const import USE_CONFIG_TIMEOUT
+from metagpt.logs import logger
+from metagpt.utils.cost_manager import CostManager
+from metagpt.utils.token_counter import count_input_tokens, count_output_tokens, get_max_completion_tokens
+from metagpt.utils.exceptions import handle_exception
+
+@register_provider(LLMType.TOGETHER)
+class TogetherLLM(BaseLLM):
```
Comment (better629) [inline metagpt/provider/together.py:15]: Notice that together has openai api compatibility inferface `https://docs.together.ai/docs/openai-api-compatibility`. So it seems that no need to create a new provider here.

**Comment by geekan** [comment]
We know that its api is compatible with the openai api, so in theory no new class is needed. But if I'm wrong, reopen it.

**Pull request closed without merging**